cmd/config-manager/main.go

```diff
@@ -165,9 +191,73 @@ func updateContainerdConfig(config map[string]interface{}) map[string]interface{
 	}
 
 	nri["disable"] = false
+
+	cfg.updateContainerdConfig(config)
+
 	return config
 }
 
+func (cfg *nriConfig) check() error {
+	switch {
+	case cfg.registrationTimeout == "" && cfg.requestTimeout == "":
+		return nil
+	case cfg.registrationTimeout != "" && cfg.requestTimeout == "":
+		return fmt.Errorf("NRI plugin registration timeout set without request timeout")
+	case cfg.registrationTimeout == "" && cfg.requestTimeout != "":
+		return fmt.Errorf("NRI plugin request timeout set without registration timeout")
+	}
+
+	register, err := time.ParseDuration(cfg.registrationTimeout)
+	if err != nil {
+		return fmt.Errorf("invalid plugin registration timeout: %w", err)
+	}
+	request, err := time.ParseDuration(cfg.requestTimeout)
+	if err != nil {
+		return fmt.Errorf("invalid plugin request timeout: %w", err)
+	}
+
+	if register <= request {
+		return fmt.Errorf("NRI plugin registration timeout (%s) must be > request timeout (%s)",
+			register, request)
+	}
+
+	return nil
+}
+
+func (cfg *nriConfig) writeCrioConfig(f *os.File) error {
+	const (
+		registrationTimeout = "nri_plugin_registration_timeout"
+		requestTimeout      = "nri_plugin_request_timeout"
+	)
+	if cfg.registrationTimeout != "" {
+		key, value := registrationTimeout, "\""+cfg.registrationTimeout+"\""
+		if _, err := f.WriteString(key + " = " + value + "\n"); err != nil {
+			return err
+		}
+	}
+	if cfg.requestTimeout != "" {
+		key, value := requestTimeout, "\""+cfg.requestTimeout+"\""
+		if _, err := f.WriteString(key + " = " + value + "\n"); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cfg *nriConfig) updateContainerdConfig(tomlCfg map[string]interface{}) {
+	const (
+		registrationTimeout = "plugin_registration_timeout"
+		requestTimeout      = "plugin_request_timeout"
+	)
+
+	if cfg.registrationTimeout != "" {
```

Comment:
If a slow NRI plugin is installed first with registrationTimeout 30s, and then a fast plugin with registrationTimeout 2s, we currently replace needed 30s timeout with 2s. Therefore I'm tempted to suggest always using max(currentRegistrationTimeout, newRegistrationTimeout).

An obvious downside of this suggestion is that if a timeout is too long, one cannot be reset to original unless we would have some kind of "force reset" bit telling this should be done.

I'd still prefer playing safe with max(). Philosophically, NRI plugins contain logic that is likely to be crucial for expected (correct) behavior of the whole node. Therefore, if one has to choose whether to give too much or too little time for a plugin to register, it's safer to give too much. If plugin needs only a little, we lose nothing. But we can loose a lot by living on the edge where plugins may be kicked out just because of fluctuation in node load.

Reply:
> If a slow NRI plugin is installed first with registrationTimeout 30s, and then a fast plugin with registrationTimeout 2s, we currently replace needed 30s timeout with 2s. Therefore I'm tempted to suggest always using max(currentRegistrationTimeout, newRegistrationTimeout).

I have a few reservations with that.

1. I think we should be careful not to give the impression that the primary means to manage the NRI-related bits in the runtime configuration would be injecting config-manager as an init container.
2. The semantics of these timeouts patched in via the config-manager should not be considered 'this plugin wants a registration timeout X'. The semantics are statedly 'I want the runtime to be reconfigured to enable NRI, and reconfigure the global common NRI timeouts for plugin registration and request handling'. We try to emphasize this by calling the timeouts on the user-visible interface (Helm chart values)  `nri.runtime.config.plugin{Registration,Request}Timeout` and not `nri.plugin.{registration,request}Timeout`.

On the practical front, there are shortcomings in the current config-manager implementation that would make it a bit laborous to add. Currently config-manager only reads (and parses to toml) the full configuration of containerd, but not CRI-O.

> 
> An obvious downside of this suggestion is that if a timeout is too long, one cannot be reset to original unless we would have some kind of "force reset" bit telling this should be done.
> 
> I'd still prefer playing safe with max(). Philosophically, NRI plugins contain logic that is likely to be crucial for expected (correct) behavior of the whole node. Therefore, if one has to choose whether to give too much or too little time for a plugin to register, it's safer to give too much. If plugin needs only a little, we lose nothing. But we can loose a lot by living on the edge where plugins may be kicked out just because of fluctuation in node load.

Again, I've always considered patching the configuration using config-manager a stop-gap solution to lower that bar for entry. To allow folks to deploy and play around with NRI plugins in their clusters with relative ease, even though the runtimes default to NRI being off. @kad ?

Reply:
I don't have strong preferences here. In theory, we shouldn't have situation when two instances of helm chart with different plugins are trying to (re-)configure runtime. The second one should detect that if NRI is already enabled/configured then just exit. Not sure with error code or with 0... both options have their own + and -.